### PR TITLE
✨(backends) add Elasticsearch database backend

### LINKED ISSUE #16: Add push command
### Purpose

Ralph should be able to write to storage backends.

### Proposal

Implement a generic `push` command that takes a storage backend (and its configuration) as arguments and streams the standard input to this backend.

### LINKED ISSUE #17: Add ElasticSearch backend
### Purpose

Ralph should be able to read/write from ES indexes.

### Proposal

Implement a new storage backend with read and write features to ES clusters.

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

Ralph should be able to publish or get statements from database backends; our first implementation of such support should be Elasticsearch since its used in production to mine collected xAPI statements.

## Proposal

- [x] add `ESDatabase` backend
- [x] adapt the `fetch` command to support both database & storage backends ; to get all database records (from an index, a collection, a table depending on supported backend)
- [x] adapt the `push` command to support both database & storage backends
- [x] add an `elasticsearch` service in the CI to run tests

Fix #16 
Fix #17

### CODE REVIEW

**Comment by jmaupetit** [comment]
This PR is blocked while #25 has not been merged.

File: src/ralph/backends/database/es.py
Comment (SergioSim) [inline src/ralph/backends/database/es.py:27]: self._hosts is not used ?
```suggestion

```

File: bin/es
Comment (madmatah) [inline bin/es:36]: Another alternative here, you can remove the `$now` variable and use the following :

```suggestion
    at=$(date -d "now -${i}min" +"%s.%N000")
```

The only difference is that `now` will not be a fixed value and will be relative to the time when the command is executed. In your use case i don't think it's a problem.

**Review by madmatah — APPROVED** [review]

File: Makefile
Comment (lunika) [inline Makefile:94]: You could also call dockerize to wait ES to be ready ?

File: src/ralph/cli.py
Comment (lunika) [inline src/ralph/cli.py:132]: the typo is intended ? `class` is a reserved name but you can maybe found something more relevant ?

File: src/ralph/utils.py
Comment (lunika) [inline src/ralph/utils.py:36]: same here, `klass` is not relevant

File: src/ralph/backends/database/es.py
Comment (lunika) [inline src/ralph/backends/database/es.py:18]: If you want to explore more OOP, you can declare a `@property` method returning this value ? You are sure this value will never change and you can also declare it as an abstract method in the abstract class. WDYT ?

File: tests/backends/database/test_es.py
Comment (lunika) [inline tests/backends/database/test_es.py:41]: IMO a private method should not be test. You will test it through the public one using it.

**Review by Sifflex — APPROVED** [review]

File: src/ralph/backends/database/es.py
Comment (jmaupetit) [reply src/ralph/backends/database/es.py:18]: It's a class attribute. It is not supposed to change.

File: tests/backends/database/test_es.py
Comment (jmaupetit) [reply tests/backends/database/test_es.py:41]: Isolated a piece of code in a private method is to make it easily reusable AND testable. I'm not sure it shouldn't be tested.

File: src/ralph/cli.py
Comment (jmaupetit) [reply src/ralph/cli.py:132]: This is not a typo :sweat_smile: I'll make it more explicit.

File: .circleci/config.yml
Comment (jmaupetit) [reply .circleci/config.yml:189]: It is a shell built-in variable as the `PATH` variable for example.

File: bin/es
Comment (jmaupetit) [reply bin/es:36]: Wow didn't know about this pattern, thanks.

File: src/ralph/backends/database/es.py
Comment (jmaupetit) [reply src/ralph/backends/database/es.py:27]: You are right, I should use it to instantiate the Elasticsearch client.

File: bin/es
Comment (jmaupetit) [reply bin/es:31]: I might be wrong but this add zeros at the end of the timestamp instead of multiplying it by 10^3.

Comment (madmatah) [reply bin/es:31]: You're right, I misinterpreted your code.

**Comment by jmaupetit** [comment]
Thank you all for your reviews :pray: @lunika waiting for your response/approval before merging this.

File: src/ralph/backends/database/es.py
Comment (lunika) [reply src/ralph/backends/database/es.py:18]: well maybe but nothing prevent you to a change of this attribute.

File: tests/backends/database/test_es.py
Comment (lunika) [reply tests/backends/database/test_es.py:41]: > solated a piece of code in a private method is to make it easily reusable and testable
I disagree, why not implementing a public method in that case ? 

You are testing twice.

Comment (lunika) [reply tests/backends/database/test_es.py:41]: A unit test is a client of the object you are testing. Never a client will access your private method, if python implement real private method you will see how difficult it is to test a private method.

Comment (jmaupetit) [reply tests/backends/database/test_es.py:41]: Ok, I've made this method public instead.

**Review by lunika — APPROVED** [review]

**Pull request merged**
